File: app/src/main/java/com/example/lovereminder/DiarySearchingActivity.java
package com.example.lovereminder;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.ImageView;
import android.widget.Toast;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class DiarySearchingActivity extends AppCompatActivity {

    private androidx.appcompat.widget.SearchView svDiary;
    private RecyclerView rvDiaries;

    private List<Diary> diaries;
    private String mQuery;
    private DiaryAdapter mAdapter;
    private DiaryDao mDiaryDao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_diary_searching);

        mDiaryDao = AppDatabase.getInstance(this).getDiaryDao();

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);

        rvDiaries = (RecyclerView) findViewById(R.id.rcv_diaries);
        mAdapter = new DiaryAdapter(this);
        rvDiaries.setAdapter(mAdapter);
        rvDiaries.setLayoutManager(new LinearLayoutManager(this));

        handleIntent(getIntent());
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            mQuery = intent.getStringExtra(SearchManager.QUERY);
            Toast.makeText(this, mQuery, Toast.LENGTH_SHORT).show();
            //use the query to search data
            mDiaryDao.getDiariesBasedOnString("*" + mQuery + "*")
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(searchedDiaries -> {
                        mAdapter.submitList(searchedDiaries);
                        mAdapter.setListener(new DiaryAdapter.Listener() {
                            @Override
                            public void onDiaryLongClickListener(Diary diary, View v) {

                            }

                            @Override
                            public void onDiaryClickListener(Diary diary) {
                                Intent intent = new Intent(DiarySearchingActivity.this, DiaryActivity.class);
                                intent.putExtra("id", diary.getId());
                                startActivity(intent);
                            }
                        });
                    });
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search_diary, menu);
        // Associate searchable configuration with the SearchView
        SearchManager searchManager =
                (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        MenuItem searchMenuItem = menu.findItem(R.id.action_search_diary);
        svDiary = (SearchView) searchMenuItem.getActionView();
        svDiary.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        svDiary.setMaxWidth(Integer.MAX_VALUE);
        //update the query in search view
        svDiary.setImeOptions(EditorInfo.IME_ACTION_SEARCH);
        //remove search icon
        svDiary.setIconifiedByDefault(false);
        svDiary.setIconified(false); // this is important to hide the search hint icon
        ImageView icon = (ImageView) svDiary.findViewById(androidx.appcompat.R.id.search_mag_icon);
//        icon.setVisibility(View.GONE);
        icon.setImageDrawable(null);
        //automatically expand the action view
        searchMenuItem.expandActionView();
        svDiary.setQuery(mQuery, false);
        return super.onCreateOptionsMenu(menu);
    }
}

File: app/src/main/java/com/phucnguyen/lovereminder/ui/activity/DiarySearchingActivity.kt
package com.phucnguyen.lovereminder.ui.activity

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.phucnguyen.lovereminder.R
import com.phucnguyen.lovereminder.database.AppDatabase.Companion.getInstance
import com.phucnguyen.lovereminder.database.DiaryDao
import com.phucnguyen.lovereminder.model.Diary
import com.phucnguyen.lovereminder.ui.adapter.DiaryAdapter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class DiarySearchingActivity : BaseActivity() {
    private var svDiary: SearchView? = null
    private var rvDiaries: RecyclerView? = null
    private val diaries: List<Diary>? = null
    private var mQuery: String? = null
    private var mAdapter: DiaryAdapter? = null
    private var mDiaryDao: DiaryDao? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_diary_searching)
        mDiaryDao = getInstance(this).diaryDao
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayShowTitleEnabled(false)
        rvDiaries = findViewById<View>(R.id.rcv_diaries) as RecyclerView
        mAdapter = DiaryAdapter(this)
        rvDiaries!!.adapter = mAdapter
        rvDiaries!!.layoutManager = LinearLayoutManager(this)
        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            mQuery = intent.getStringExtra(SearchManager.QUERY)
            Toast.makeText(this, mQuery, Toast.LENGTH_SHORT).show()
            //use the query to search data
            mDiaryDao!!.getDiariesBasedOnString("*$mQuery*")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { searchedDiaries: List<Diary?> ->
                    mAdapter!!.submitList(searchedDiaries)
                    mAdapter!!.setListener(object : DiaryAdapter.Listener {
                        override fun onDiaryLongClickListener(diary: Diary, v: View) {}

                        override fun onDiaryClickListener(diary: Diary) {
                            val intent =
                                Intent(this@DiarySearchingActivity, DiaryActivity::class.java)
                            intent.putExtra("id", diary?.id)
                            startActivity(intent)
                        }
                    })
                }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search_diary, menu)
        // Associate searchable configuration with the SearchView
        val searchManager = getSystemService(SEARCH_SERVICE) as SearchManager
        val searchMenuItem = menu.findItem(R.id.action_search_diary)
        svDiary = searchMenuItem.actionView as SearchView?
        svDiary!!.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        svDiary!!.maxWidth = Int.MAX_VALUE
        //update the query in search view
        svDiary!!.imeOptions = EditorInfo.IME_ACTION_SEARCH
        //remove search icon
        svDiary!!.setIconifiedByDefault(false)
        svDiary!!.isIconified = false // this is important to hide the search hint icon
        val icon =
            svDiary!!.findViewById<View>(androidx.appcompat.R.id.search_mag_icon) as ImageView
        //        icon.setVisibility(View.GONE);
        icon.setImageDrawable(null)
        //automatically expand the action view
        searchMenuItem.expandActionView()
        svDiary!!.setQuery(mQuery, false)
        return super.onCreateOptionsMenu(menu)
    }
}